Zoom via mouse wheel: keep the position under the cursor in place

smurfix opened the pull request:
#### Brief overview of PR changes/additions
Mouse wheel zooming (2D view) previously kept the current map center in place. Thus there's no way to move around in the map other than using the left/right/up/down buttons, which is painful.

Now the position under the cursor is kept in place. Thus you can easily and quickly look at some other map region -- zoom out, move mouse to region of interest, zoom in.

#### Motivation for adding to Mudlet
Obvious ;-)

smurfix committed:
Zoom centered on the mouse position

Zooming via the scroll wheel now keeps the position under the curser in
place.

vadi2 reviewed (commented):
Thanks for this!

https://doc.qt.io/qt-5/qwheelevent.html#position is in 5.14 only - while we maintain compatibility with 5.11. Could you put in #ifdef for the two versions? `#if (QT_VERSION >= QT_VERSION_CHECK(5, 14, 0))`

smurfix committed:
QT <5.14 compatibility

smurfix commented:
Let's see if this works.

smurfix committed:
Use an exponential factor when zooming

This makes the zoom experience a lot smoother: the step size no longer
depends on how close you're zoomed in.

smurfix commented:
Further improvement: change the zoom level by a factor instead of by a constant.
This makes the whole thing a lot smoother: no super large steps when zoomed in, no "dead" feeling when zoomed far out.

SlySven commented on src/T2DMap.cpp:
```diff
@@ -4611,13 +4611,42 @@ void T2DMap::wheelEvent(QWheelEvent* e)
+        qreal oldZoom = xyzoom;
+        xyzoom = qMax(3.0, xyzoom * pow(1.07, delta));
+
         if (oldZoom != xyzoom) {
```
It is bad/wrong to test for equality/inequality on *floating-point* numbers - consider `qFuzzyCompare(...)` I think it is - instead!

smurfix replied:
See other comment; while I can certainly change that, in this case it's at worst slightly suboptimal when you touch the 3.0 barrier.

SlySven commented on src/T2DMap.cpp:
```diff
@@ -4611,13 +4611,42 @@ void T2DMap::wheelEvent(QWheelEvent* e)
     if (e->modifiers() & Qt::ControlModifier) { // Increase rate 10-fold if control key down - it makes scrolling through
```
Perhaps this should be 5-fold now!

smurfix replied:
Actually 5.7: `(F^5 - 1) / F - 1)` with `F=1.07`.
I'll remove that number.

SlySven commented on src/T2DMap.cpp:
```diff
@@ -4611,13 +4611,42 @@ void T2DMap::wheelEvent(QWheelEvent* e)
+        xyzoom = qMax(3.0, xyzoom * pow(1.07, delta));
+
         if (oldZoom != xyzoom) {
+            const float widgetWidth = width();
```
Both `width()` and `height()` return `int`s - do you need `static_cast<float>(`...`)` to suppress type conversion warnings?

smurfix replied:
There's no such warning even with gcc's `-Wall`.

SlySven commented on src/T2DMap.cpp:
```diff
@@ -4611,13 +4611,42 @@ void T2DMap::wheelEvent(QWheelEvent* e)
         if (oldZoom != xyzoom) {
+            const float widgetWidth = width();
+            const float widgetHeight = height();
+            float xs = 1.0;
```
Don't you need a `f` suffix to the `1.0` to register it as a literal `float` constant rather than a `double` (the default) one.

smurfix replied:
It's a constant with no loss of precision when converted to a `float`, so the compiler does the right thing. No conversion warning.

SlySven commented on src/T2DMap.cpp:
```diff
@@ -4628,10 +4657,10 @@ void T2DMap::wheelEvent(QWheelEvent* e)
-    xyzoom = qMax(3, zoom);
+    qreal oldZoom = xyzoom;
+    xyzoom = qMax(3.0, zoom);
     if (oldZoom != xyzoom) {
```
Another equality/inequality test on floating point numbers!

smurfix replied:
Well, yes. However, float comparisons are dangerous only if you have fractional values that may end up close to each other. Here we're using 3.0 as a hard boundary (exact value), otherwise `oldZoom` and `xyzoom` will always be either significantly different or identical (multiplying with 1.0 never changes a float), and even if zooming back in ends up at 3.000001 the single additional step down to 3.0 doesn't hurt anybody.

Thus I didn't bother to look up whatever float-almost-equality method Qt has. Though I can certainly add it ;-)

SlySven reviewed (commented):
What about the corresponding code for the 3D mapper?

smurfix commented:
I haven't yet looked at the 3D mapper. A brief test, right now, shows that it doesn't suffer from the issues the topmost change fixes; it has other problems, such as scrolling away beyond infinity, no way to rotate the view with the mouse (like in every other 3D viewer coded during the last ten years or so), view rotation scrollbars that are much too granular, and whatnot. So no, somebody else shall tackle its set of problems.

SlySven reviewed (approved):
:+1: Actually after trying this, it does do what it says on the tin...

vadi2 commented:
Sorry for the delay in this! Thanks for adding it, really nice.
